---
title: "GSVA Notebook"
output: html_notebook
---

> https://bioconductor.org/packages/release/bioc/manuals/GSVA/man/GSVA.pdf

```{r}
library(GSVA)
geneSets <- list(set1=as.character(1:4), set2=as.character(4:10))
unlist(geneSets)
computeGeneSetsOverlap(geneSets, unique(unlist(geneSets)))

filterGeneSets(geneSets, min.sz = 5)
```
```{r}
library(limma)

p <- 10
n <- 30
nGrp1 <- 15
nGrp2 <- n - nGrp1

geneSets <- list(
  set1 = paste('g', 1:3, sep = ""),
  set2 = paste('g', 4:6, sep = ""),
  set3 = paste('g', 7:10, sep = "")
)


y <- matrix(rnorm(n * p), nrow = p, ncol = n, dimnames = list(paste('g', 1:p, sep = ""), paste('s', 1:n, sep = "")))

# nGrp1+1 to n express at higher levels
y[geneSets$set1, (nGrp1 + 1):n] <- y[geneSets$set1, (nGrp1 + 1):n] + 2

# build design matrix
design <- cbind(sampleGroup1 = 1, sampleGroup2vs1 = c(rep(0, nGrp1), rep(1, nGrp2)))

# fit linear model
fit <- lmFit(y, design)

# estimate moderated t-statistics

fit <- eBayes(fit)

# genes in set1 are differentially expressed 

topTable(fit, coef = 'sampleGroup2vs1')

# estimate GSVA enrichment scores for the three sets

gsva_es <- gsva(y, geneSets, mx.diff = 1)

# fit the same linear model now to the GSVA enrichment scores

fit <- lmFit(gsva_es, design = design)
fit <- eBayes(fit)

topTable(fit, coef = "sampleGroup2vs1")

```

```{r}

p <- 10
n <- 30
nGrp1 <- 15
nGrp2 <- n - nGrp1

geneSets <- list(
  set1 = paste('g', 1:3, sep = ""),
  set2 = paste('g', 4:6, sep = ""),
  set3 = paste('g', 7:10, sep = "")
)


y <- matrix(rnorm(n * p), nrow = p, ncol = n, dimnames = list(paste('g', 1:p, sep = ""), paste('s', 1:n, sep = "")))

gsva_es <- gsva(y, geneSets, mx.diff = 1)

expr <- y
gset.idx.list <- geneSets
mx.diff = 1
method <- 'gsva'
kcdf <- 'Gaussian'

sdGenes <- apply(expr, 1, sd)
if (any(sdGenes == 0) || any(is.na(sdGenes))) { 
  warning(sum(sdGenes == 0 | is.na(sdGenes)), " genes with constant expression values throughout the samples.")
  if (method != 'ssgsea') {
    warning("Since argument method != \"ssgsea\", genes with constant expression values are discarded.")
    expr <- expr[sdGenes > 0 & !is.na(sdGenes),, drop = FALSE]
  }
}
if (nrow(expr) < 2) stop("Less than two genes in the input expression data matrix\n")
# map gene indentifiers of the gene sets to the features in the matrix
mapped.gset.idx.list <- gset.idx.list

mapped.gset.idx.list <- lapply(
  gset.idx.list,
  function(x, y) na.omit(match(x, y)),
  rownames(expr)
)
min.sz <- 1
max.sz <- Inf


mapped.gset.idx.list <- filterGeneSets(
  mapped.gset.idx.list,
  min.sz = max(1, min.sz),
  max.sz = max.sz
  )

if (!missing(kcdf)) {
    if (kcdf == "Gaussian") {
      rnaseq <- FALSE
      kernel <- TRUE
    } else if (kcdf == "Poisson") {
      rnaseq <- TRUE
      kernel <- TRUE
    } else
      kernel <- FALSE
}



```


